### DISCUSSION TITLE
Software applications installations inventory management

### DISCUSSION THREAD

**Discussion by adrianandreias** [netbox-community/netbox#10238] [Q&A]
Hello,

I've noticed Netbox is more oriented to physical inventory and also has the concept of VM.

Is there a solution to keep track of server software applications installations (on own hardware or not), like I have these 2 WordPress installations, this GitLab etc.

Many times there is a 1-to-1 relation between VM or physical server and server app, but not always. E.g. I can have 3 WordPress installs on one physical server.

Can be done with Netbox, if not do can you please suggest some other tool?

Thank you.

**Comment by richbibby** [comment]
Hi, 

you can use the 'Services' feature to map applications to VM's or devices. from the docs: https://docs.netbox.dev/en/stable/models/ipam/service/ or for example: https://demo.netbox.dev/ipam/services/

Admittedly it's not tracking the number of software installations as such....

Rich

**Comment by adrianandreias** [comment]
Oh, that's really useful. And I see that two services may be on the same VM, same TCP port and same IP. Useful for virtual hosting.

I think Netbox does not have an agent. What would be a good way to script automatic app version retrieval to update it a custom field in Netbox? So that you can see in the list of services the actual versions of the software applications/services.

Thanks.

**Reply by richbibby** [reply]
A custom script would probably be the best solution for that: https://docs.netbox.dev/en/stable/customization/custom-scripts/

Outside of that you could write an externally hosted script to populate the custom field for 'version' via the NetBox REST API

**Reply by candlerb** [reply]
But remember that goes against the philosophy of Netbox - as a manually curated source of truth, not a portal on network state.

If adding or removing an item in Netbox actually *caused* that piece of software to be installed or removed or the target system by triggering some automated process - that would be more in the spirit.

**Reply by adrianandreias** [reply]
> But remember that goes against the philosophy of Netbox - as a manually curated source of truth, not a portal on network state.
> 
> If adding or removing an item in Netbox actually _caused_ that piece of software to be installed or removed or the target system by triggering some automated process - that would be more in the spirit.

I am new to Netbox. At first sight it fits very good our requirements to track hardware, racks, locations, IPs etc.

On the other hand, you want minimize the number of apps you use. That's why I am trying to squeeze internal server app inventory as well.

Considering this, I don't fully get the idea. You're saying that data should be (mostly) added manually, not retrieved automatically (e.g. version populated manually goes against), right? And Netbox goal is also to change stuff on network/devices triggered by changed in Netbox? I did not know this last part.

And now that I am going over ideas here, maybe Zabbix would be more appropriate to have a full internal app inventory. It already has an agent and basic data like version can be retrieved.

What are you guys using for sort of apps inventory? Just a full list and basic info, like version.

**Reply by richbibby** [reply]
Yes, that's a very good point!

**Reply by candlerb** [reply]
> Considering this, I don't fully get the idea. You're saying that data should be (mostly) added manually, not retrieved automatically (e.g. version populated manually goes against), right?

Yes. See:
* https://docs.netbox.dev/en/stable/introduction/#serve-as-a-source-of-truth
* https://docs.netbox.dev/en/stable/introduction/#what-netbox-is-not

> And Netbox goal is also to change stuff on network/devices triggered by changed in Netbox? I did not know this last part.

Yes. Any automation needs a trusted source of information of the form "what state is desired".

> And now that I am going over ideas here, maybe Zabbix would be more appropriate to have a full internal app inventory. It already has an agent and basic data like version can be retrieved.
> 
> What are you guys using for sort of apps inventory? Just a full list and basic info, like version.

For me: Prometheus with node_exporter / windows_exporter / snmp_exporter and Grafana dashboards is my preferred stack for collecting information from running systems - down to the level of serial numbers of hard drives (in [smartmon](https://github.com/prometheus-community/node-exporter-textfile-collector-scripts/blob/master/smartmon.sh) information).  [This script](https://github.com/prometheus-community/node-exporter-textfile-collector-scripts/blob/master/apt_info.py) tells you the number of package updates outstanding. If you want a list of individual packages, it's pretty easy to add.

Where the Netbox integration comes in is that I use Netbox to *configure* Prometheus: that is, to create the list of targets to scrape.  Adding a new device in Netbox causes it to be added to monitoring.  Therefore, the status "Active" or "Failed" in Netbox is not a reflection of what the monitoring says.  Rather: if a device goes down, and I decide it's going to be down for a while and I shouldn't be trying to scrape it for the time being, I manually change the state from "Active" to "Failed" in Netbox.  Therefore, Netbox tells the monitoring system what to do; the monitoring system does not automatically update data in Netbox.

There are other tools: e.g. Netdisco is helpful for answering questions like "where is this IP address / MAC address plugged into my network *right now*"?

**Reply by adrianandreias** [reply]
Good ideas. Thank you. Zabbix or Graphana + Prometheus + node_exporter seem to be way better for what I want.

**Reply by danie-dejager** [reply]
@adrianandreias Adding a new host to Zabbix when added to Netbox with an active state should be easy enough but you'd also need Ansible or similar to install and configure the Zabbix agent on the host itself.
